File: trees/maximum-depth-of-binary-tree.js
// https://leetcode.com/problems/maximum-depth-of-binary-tree/description/

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    maxDepth(root) {
        if (root === null) return 0; // Base case: empty tree

        // Recursively get depth of left and right subtrees
        const leftDepth = this.maxDepth(root.left);
        const rightDepth = this.maxDepth(root.right);

        // The depth of the tree = 1 (for root) + max of left and right depths
        return 1 + Math.max(leftDepth, rightDepth);
    }
}

File: python/trees/maximum_depth_of_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# we start off with the root of the binary tree
# we end with a number data type that represents the max depth
# root = [1,2,3,null,null,4] => 3
# implement DFS, traverse right and left side, add 1 and return the max

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        # base case
        if not root:
            return 0

        # recursive case
        right = self.maxDepth(root.right)
        right += 1
        left = self.maxDepth(root.left)
        left += 1

        return max(right, left)
        